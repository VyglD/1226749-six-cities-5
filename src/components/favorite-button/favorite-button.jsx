import React from "react";
import {connect} from "react-redux";
import ActionCreator from "../../store/root-actions";
import {
  offerType,
  offerIdsType,
  functionType,
  boolType,
  favoriteBtnStyleType,
} from "../../types";
import {Path} from "../../const";
import {getArraysDifference} from "../../util";

class FavoriteButton extends React.Component {
  constructor(props) {
    super(props);

    this.handleFavoriteClick = this.handleFavoriteClick.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    const {favoriteOfferIds, offer} = this.props;

    return getArraysDifference(
        nextProps.favoriteOfferIds,
        favoriteOfferIds
    ).includes(offer.id);
  }

  handleFavoriteClick() {
    const {
      favoriteOfferIds,
      isLogin,
      offer,
      deleteFavorite,
      addFavorite,
      redirectTo,
    } = this.props;

    if (isLogin) {
      if (favoriteOfferIds.includes(offer.id)) {
        deleteFavorite(offer.id);
      } else {
        addFavorite(offer.id);
      }
    } else {
      redirectTo(Path.LOGIN);
    }
  }

  render() {
    const {
      favoriteBtnStyle: {
        btnClassName = `place-card__bookmark-button`,
        btnActiveClassName = `place-card__bookmark-button--active`,
        iconClassName = `place-card__bookmark-icon`,
        iconWidth = 18,
        iconHeight = 19,
      } = {},
      favoriteOfferIds,
      offer,
    } = this.props;

    return (
      <button
        className={`button ${btnClassName} ${
          favoriteOfferIds.includes(offer.id) ? btnActiveClassName : ``
        }`}
        type="button"
        onClick={this.handleFavoriteClick}
      >
        <svg className={iconClassName} width={iconWidth} height={iconHeight}>
          <use xlinkHref="#icon-bookmark"></use>
        </svg>
        <span className="visually-hidden">To bookmarks</span>
      </button>
    );
  }
}

FavoriteButton.propTypes = {
  offer: offerType,
  favoriteOfferIds: offerIdsType,
  isLogin: boolType,
  deleteFavorite: functionType,
  addFavorite: functionType,
  redirectTo: functionType,
  favoriteBtnStyle: favoriteBtnStyleType,
};

const mapStateToProps = ({USER, FAVORITES}) => ({
  isLogin: USER.isLogin,
  favoriteOfferIds: FAVORITES.favoriteOfferIds,
});

const mapDispatchToProps = (dispatch) => ({
  deleteFavorite(offerId) {
    dispatch(ActionCreator.deleteFavorite(offerId));
  },
  addFavorite(offerId) {
    dispatch(ActionCreator.addFavorite(offerId));
  },
  redirectTo(url) {
    dispatch(ActionCreator.redirectTo(url));
  }
});

export {FavoriteButton};
export default connect(mapStateToProps, mapDispatchToProps)(FavoriteButton);
