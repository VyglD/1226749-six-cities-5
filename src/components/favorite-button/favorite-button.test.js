import React from "react";
import renderer from "react-test-renderer";
import {FavoriteButton} from "./favorite-button";
import {
  mockOffers,
  mockOffersFavoriteIds,
  mockFunction
} from "../../mocks/mock-data";

const FAVORITE_BTN_STYLE = {
  btnClassName: `property__bookmark-button`,
  btnActiveClassName: `property__bookmark-button--active`,
  iconClassName: `property__bookmark-icon`,
  iconWidth: 31,
  iconHeight: 33,
};

describe(`Render correctly component FavoriteButton`, () => {
  it(`Render FavoriteButton`, () => {
    const component = renderer.create(
        <FavoriteButton
          offer={mockOffers[3]}
          favoriteIds={mockOffersFavoriteIds}
          addFavorite={mockFunction}
          deleteFavorite={mockFunction}
          favoriteBtnStyle={FAVORITE_BTN_STYLE}
        />
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  it(`Render FavoriteButton of favorite offer`, () => {
    const component = renderer.create(
        <FavoriteButton
          offer={mockOffers[0]}
          favoriteIds={mockOffersFavoriteIds}
          addFavorite={mockFunction}
          deleteFavorite={mockFunction}
          favoriteBtnStyle={FAVORITE_BTN_STYLE}
        />
    ).toJSON();

    expect(component).toMatchSnapshot();
  });
});
