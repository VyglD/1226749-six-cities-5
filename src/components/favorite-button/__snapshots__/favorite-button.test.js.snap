// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render correctly component FavoriteButton Render FavoriteButton 1`] = `
<button
  className="button place-card__bookmark-button "
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={19}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;

exports[`Render correctly component FavoriteButton Render FavoriteButton of favorite offer 1`] = `
<button
  className="button place-card__bookmark-button place-card__bookmark-button--active"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={19}
    width={18}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;

exports[`Render correctly component FavoriteButton Render FavoriteButton with custom button style 1`] = `
<button
  className="button property__bookmark-button property__bookmark-button--active"
  onClick={[Function]}
  type="button"
>
  <svg
    className="property__bookmark-icon"
    height={33}
    width={31}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;
